scala/bigtable-shc/src/main/scala/com/example/bigtable/spark/shc/BigtableSource.scala

```diff
@@ -16,75 +16,85 @@
 
 package com.example.bigtable.spark.shc
 
-import org.apache.spark.sql.execution.datasources.hbase._
-import org.apache.spark.sql.SparkSession
+object BigtableSource extends App {
 
-case class BigtableRecord(
-                        col0: String,
-                        col1: Boolean,
-                        col2: Double,
-                        col3: Int)
+  val tableName = args(0)
+  val cat =
+    s"""{
+       |"table":{"namespace":"default", "name":"$tableName", "tableCoder":"PrimitiveType"},
+       |"rowkey":"key",
+       |"columns":{
+       |"col0":{"cf":"rowkey", "col":"key", "type":"string"},
+       |"col1":{"cf":"cf1", "col":"col1", "type":"boolean"},
+       |"col2":{"cf":"cf2", "col":"col2", "type":"double"},
+       |"col3":{"cf":"cf3", "col":"col3", "type":"int"}
+       |}
+       |}""".stripMargin
+  val appName = getClass.getSimpleName.replace("$", "")
 
-object BigtableRecord {
-  def apply(i: Int): BigtableRecord = {
-    val s = s"""row${"%03d".format(i)}"""
-    BigtableRecord(s,
-      i % 2 == 0,
-      i.toDouble,
-      i)
-  }
-}
+  import org.apache.spark.sql.SparkSession
 
-object BigtableSource {
+  val spark = SparkSession
+    .builder
+    .appName(appName)
+    .getOrCreate
 
-  def main(args: Array[String]) {
-    val tableName = args(0)
+  import spark.implicits._
```

Comment:
imports at the top?

Reply:
That one is not possible since it imports (brings values into the current namespace) from the object `spark` that was created a few lines earlier. `spark` is an instance not a package and the import is to have Scala implicits in scope. That's how Scala implicits work.

Reply:
Got it.